Support large lists of fullnames for `hide` and `get_info`

voussoir opened the pull request:
What I did:

- New helper function `chunk_sequence` converts a long list into multiple smaller lists of a certain length.
- Changed `r.hide` and `r.get_info` to use `chunk_sequence` so that the user can input any number of fullnames and have all the requests processed automatically. This involves making "param groups" for each request and adding the server response into a total final list.
- `url = True  # Enable returning a list` was confusing, so I've made `return_list` its own variable.
- Added comment stickying to the changelog which was forgotten last time.

Feedback wanted:
- My primary concern at the moment is the return value of `r.hide`. Previously it was just the server response, which as far as I know is always an empty dictionary as long as the request is successful. Now, I'm returning the list of these empty dictionaries. Does this seem like the best route? This could confuse programs which expect the dictionary, but I'm kind of assuming no one uses these for anything.
- Any other methods that would benefit from this chunking?

Haven't made a pull request in a while, hopefully I still remember how to make Travis pass.

&nbsp;

Edit: If I have to make any further commits I will change my list comprehension in `test_info_by_id` to please Travis 2.7. Otherwise I'm going to let that sit for now.

bboe commented on tests/test_unauthenticated_reddit.py:
```diff
@@ -211,8 +211,18 @@ def test_get_top(self):
+        submission = self.r.get_info(thing_id=self.link_id)
+        self.assertEqual(self.link_id, submission.fullname)
+
+        # groups
```
Can we make the group a separate test so that we can ensure the original cassette still works without modification?

bboe commented on tests/test_authenticated_reddit.py:
```diff
@@ -96,6 +96,13 @@ def test_submission_hide_and_unhide_batch(self):
 
+        submissions = list(self.r.get_subreddit('all').get_new(limit=300))
+        fullnames = [submission.fullname for submission in submissions]
+        self.r.hide(fullnames)
```
Can we make the group functionality a separate test to ensure the original cassette still works as expected?

bboe commented on praw/__init__.py:
```diff
@@ -2569,28 +2583,36 @@ class ReportMixin(AuthenticatedReddit):
+            if self.user is not None:
+                self.evict(urljoin(self.user._url,  # pylint: disable=W0212
+                                   'hidden'))
+        return responses
```
Am I correct that this changes the response type always to a list? Can we make it such that it matches the input format (e.g., list in list out, single item in, single item out) so that this isn't a breaking change?

bboe commented on praw/helpers.py:
```diff
@@ -356,6 +356,28 @@ def b36_id(item):
             time.sleep(sleep_time)
 
 
+def chunk_sequence(sequence, chunk_length, allow_incomplete=True):
```
Do you think it would be worthwhile to have unit tests to ensure this function works as expected?

bboe commented:
I quickly scanned over the changes and made a few comments. Awesome work.

voussoir commented:
Thanks for the notes Bryce, I'll be glad to move the batch tests into their own unit. I'll also create `test_helpers.py` to test `chunk_sequence`.

> Am I correct that this changes the response type always to a list? Can we make it such that it matches the input format (e.g., list in list out, single item in, single item out) so that this isn't a breaking change?

This is correct. However, list-in-list-out would still be a breaking change, because previously you could pass in a list of up to 50 items and you'd only get back that single `{}` response from the server. The true non-breaking way to do this would be

- single response for individual or lists <=50;
- list of responses for more.

I'm fine adopting this approach for now as long as we do PRAW4 the way you describe.

voussoir commented:
Okay, changes made. Here's what I've done:


- Separated the new tests into their own units.
- Moved `test_submission_hide_and_unhide_batch` from `test_authenticated_reddit.py` to `test_submission.py` where the other hide test was.
- Added `test_chunk_sequence` to `test_helpers.py`
- Made `r.hide` return a singular response dictionary if the user inputted less than 50 items, as it has behaved in the past. Wrote a TODO explaining this 50 item threshold.
- Added some more asserts to pre-existing hide tests to make sure `r.hide` is returning the right number of response dicts.
- Converted `thing_id` to a list before chunking, since `chunk_sequence` doesn't work on sets, generators, etc.

Let me know if you see anything else!

bboe commented:
Looks fantastic. Thanks @voussoir. Merge and make a release if you'd like 👍

voussoir commented:
Thank you! Will do.
